File: 04_Philosophers/main.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#define NUM_PHILOSOPHERS 5
#define ITERATIONS 3

static pthread_mutex_t sticks[NUM_PHILOSOPHERS];

static void *philosopher(void *arg) {
    int id = *(int *)arg;
    free(arg);

    int left = id;
    int right = (id + 1) % NUM_PHILOSOPHERS;

    for (int i = 0; i < ITERATIONS; i++) {
        printf("Filozof %d razmislja.\n", id);
        fflush(stdout);
        usleep((10 + rand() % 50) * 1000);

        printf("Filozof %d pokusava da uzme stapice.\n", id);
        fflush(stdout);

        /* ASIMETRIČNO REŠENJE: Neparni filozofi uzimaju kontra redosledom */
        if (id % 2 != 0) {
            pthread_mutex_lock(&sticks[right]);
            pthread_mutex_lock(&sticks[left]);
        } else {
            pthread_mutex_lock(&sticks[left]);
            pthread_mutex_lock(&sticks[right]);
        }

        printf("Filozof %d je uspesno uzeo oba stapica i jede.\n", id);
        fflush(stdout);
        usleep((10 + rand() % 50) * 1000);

        /* Vraćanje štapića */
        pthread_mutex_unlock(&sticks[left]);
        pthread_mutex_unlock(&sticks[right]);

        printf("Filozof %d je vratio stapice.\n", id);
        fflush(stdout);
    }

    return NULL;
}

int main(void) {
    srand(time(NULL));
    pthread_t phils[NUM_PHILOSOPHERS];

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        pthread_mutex_init(&sticks[i], NULL);
    }

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        int *id = malloc(sizeof(int));
        *id = i;
        pthread_create(&phils[i], NULL, philosopher, id);
    }

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        pthread_join(phils[i], NULL);
    }

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        pthread_mutex_destroy(&sticks[i]);
    }

    printf("\n[USPEH] Svi filozofi su zavrsili sa jelom i razmisljanjem!\n");
    return 0;
}

File: 04_Philosophers/main.go
/*
 * Dining Philosophers - asimetricno resenje
 *
 * -----------------------------------------------------------
 * C vs Go:
 *   pthread_mutex_t sticks[N]         ->  var sticks [N]sync.Mutex
 *   pthread_mutex_lock(&sticks[i])    ->  sticks[i].Lock()
 *   pthread_mutex_unlock(&sticks[i])  ->  sticks[i].Unlock()
 * -----------------------------------------------------------
 */

package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	NUM_PHILOSOPHERS = 5
	ITERATIONS       = 3
)

var sticks [NUM_PHILOSOPHERS]sync.Mutex

func philosopher(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	left := id
	right := (id + 1) % NUM_PHILOSOPHERS

	for i := 0; i < ITERATIONS; i++ {
		fmt.Printf("Filozof %d razmislja.\n", id)
		time.Sleep(time.Duration(10+rand.Intn(50)) * time.Millisecond)

		fmt.Printf("Filozof %d pokusava da uzme stapice.\n", id)

		// ASIMETRICNO RESENJE: neparni filozofi uzimaju kontra redosledom
		if id%2 != 0 {
			sticks[right].Lock()
			sticks[left].Lock()
		} else {
			sticks[left].Lock()
			sticks[right].Lock()
		}

		fmt.Printf("Filozof %d je uspesno uzeo oba stapica i jede.\n", id)
		time.Sleep(time.Duration(10+rand.Intn(50)) * time.Millisecond)

		sticks[left].Unlock()
		sticks[right].Unlock()

		fmt.Printf("Filozof %d je vratio stapice.\n", id)
	}
}

func main() {
	var wg sync.WaitGroup

	for i := 0; i < NUM_PHILOSOPHERS; i++ {
		wg.Add(1)
		go philosopher(i, &wg)
	}

	wg.Wait()
	fmt.Println("\n[USPEH] Svi filozofi su zavrsili sa jelom i razmisljanjem!")
}
